src/refresh/RefreshTimer.ts

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+/**
+ * The backoff time is between the minimum and maximum backoff time, based on the number of attempts.
+ * An exponential backoff strategy is used, with a jitter factor to prevent clients from retrying at the same time.
+ *
+ * The backoff time is calculated as follows:
+ *  - `basic backoff time` = `MinimumBackoffInMs` * 2 ^ `attempts`, and it is no larger than the `MaximumBackoffInMs`.
+ *  - based on jitter ratio, the jittered time is between [-1, 1) * `JitterRatio` * basic backoff time.
+ *  - the final backoff time is the basic backoff time plus the jittered time.
+ *
+ * Note: the backoff time usually is no larger than the refresh interval, which is specified by the user.
+ *  - If the interval is less than the minimum backoff, the interval is used.
+ *  - If the interval is between the minimum and maximum backoff, the interval is used as the maximum backoff.
+ *  - Because of the jitter, the maximum backoff time is actually `MaximumBackoffInMs` * (1 + `JitterRatio`).
+ */
+
+const MinimumBackoffInMs = 30 * 1000; // 30s
+const MaximumBackoffInMs = 10 * 60 * 1000; // 10min
+const MaxSafeExponential = 53; // Used to avoid overflow, as Number.MAX_SAFE_INTEGER = 2^53 - 1.
+const JitterRatio = 0.25;
+
+export class RefreshTimer {
+    private _minBackoff: number = MinimumBackoffInMs;
+    private _maxBackoff: number = MaximumBackoffInMs;
+    private _failedAttempts: number = 0;
+    private _backoffEnd: number; // Timestamp
+    constructor(
+        private _interval: number
+    ) {
+        if (this._interval <= 0) {
+            throw new Error(`Refresh interval must be greater than 0. Given: ${this._interval}`);
+        }
+
+        this._backoffEnd = Date.now() + this._interval;
+    }
+
+    public canRefresh(): boolean {
+        return Date.now() >= this._backoffEnd;
+    }
+
+    public backoff(): void {
+        this._backoffEnd = Date.now() + this._calculateBackoffTime();
+        this._failedAttempts += 1;

```

Comment:
If we initialized `_failedAttempts` to 0, we should increment this before calculating the backoff time.

Reply:
current calculated time is correct. 
Then we probably need to change calculateBackoffTime, making it equal to pow(2, failedAttempts-1) * minBackoff

Reply:
Or we denote the `failedAttempts` with "failed attempts before current one"... Anyway either way is ok for me.

Reply:
I might be missing something here. Why do we need to calculate failedAttempts-1?
We don't need to calculate back off for the first attempt (2^1) because the first time we can simply back off for the min duration (just an optimisation). 
We do need to calculate back off from the second attempt onwards (2^2).